### PR TITLE
Search: allow search several projects at the same time

### PR DESCRIPTION
**Pull request by stsewd**
This implements https://github.com/readthedocs/readthedocs.org/issues/8678#issuecomment-1163780743.

- Accept a list of projects via the "project" parameter,
  each project can be in the form of `{project_slug}` or `{project_slug:version_slug}`,
  if the version isn't present, we use the default version of the project.
- It will include only the results from the versions that were explicitly requested (no subprojects)
- Every version will be validated if the user has permission over it,
  if the user doesn't have permission over one or if it doesn't exist, we just don't include results from that version.
- The old "syntax" is still available (it includes results from subprojects), and it requires using the `version` parameter.

ref https://github.com/readthedocs/readthedocs.org/issues/8678

Front logo [Front conversations](https://app.frontapp.com/open/top_5c8g7)

### CODE REVIEW

File: docs/user/server-side-search.rst [hunk 1]
```diff
@@ -188,15 +203,43 @@ Check :ref:`server-side-search:authentication and authorization` if you are usin
...
+users will only be allowed to search versions they have permissions over.
 Authentication and authorization is done using the current session,
 or any of the valid :doc:`sharing methods </commercial/sharing>`.
 
 To be able to use the user's current session you need to use the API from the domain where your docs are being served
 (``<you-docs-domain>/_/api/v2/search/``).
 This is ``https://docs.readthedocs-hosted.com/_/api/v2/search/``
 for the ``https://docs.readthedocs-hosted.com/`` project, for example.
+
+Including results from subprojects
```
[the remark this answers is not in the document]
Comment (ericholscher) [reply docs/user/server-side-search.rst:241]: Yes, this is not a useful piece of documentation. We should definitely make this nicer for users, as most shouldn't be changing how our default search experience works.

**Review by ericholscher — COMMENTED** [review]
I haven't done a full review of this code, but I have a few questions:

* How does this search interact with our existing site search? It feels like they are solving similar problems, but with pretty different approaches to URL design. Is it possible to unify them with this work?
* I would _love_ to see the ability for users to have more powerful search in the query. This is almost 100% what GitHub and other search providers are doing these days, and we already support it: https://docs.readthedocs.io/en/latest/guides/advanced-search.html#search-query-syntax -- Did we consider using the query syntax for filtering instead of different API parameters? Is it possible to still implement this in the query for the user, but modify the values going into the API? Is that what we should be doing? 

I want to understand the larger context here before I do a full review of the code.

File: docs/user/server-side-search.rst [hunk 2]
```diff
@@ -82,30 +82,38 @@ Check :ref:`server-side-search:authentication and authorization` if you are usin
...
 
-   :query q: Search query
-   :query project: Project slug
-   :query version: Version slug
-   :query page: Jump to a specific page
-   :query page_size: Limits the results per page, default is 50
+   :query string q: Search query.
+   :query string | array project: A project or list of projects to search,
+    they can be in the form of ``{project}`` or ``{project}:{version}``.
+    If no version is passed, its default version is used.
```
[the remark this answers is not in the document]
Comment (stsewd) [reply docs/user/server-side-search.rst:93]: The standard way of passing an array in query parameters is to just duplicate the same parameter, e.g: `?project=foo&project=bar`, there is an example of this just below

[the remark this answers is not in the document]
Comment (benjaoming) [reply docs/user/server-side-search.rst:93]: Great to have a specific example! Unfortunately, there are competing standards for query parameter arrays, so readers might have different experiences and expectations. Example: https://stackoverflow.com/a/9547490/405682

**Comment by ericholscher** [comment]
@stsewd Sorry I haven't gotten back to discuss this. I think this is useful work, so let's go ahead and focus on shipping it and we can think about the dashboard implications in the next round of work on this. 

> Hmm, so the site search is searching all projects, do we want to expose that in our API? I think it makes sense on .org, not sure about .com? On .com we search all projects the user has access to. Maybe keep that same behavior?

I'd like to have this be the default behavior of the dashboard search on both sites. I think it's way more useful for users, and makes both sites act the same. 

> It's possible to do it that way, we will need to parse the query and maybe check for some ambiguities. My main reason to use query parameters was so devs could define what projects to search in one endpoint, but I guess the same can be achieved if the parameters are in the query itself (q="{prepared-parameters}+{actual user query}")

Yea, I think having this functionality be exposed to users is pretty valuable. Both should work though, and we can combine them on our side? 

> No really strong opinion whether to make a new endpoint or not, the only difference will be the input, the output will be the same (or almost, just new fields will be added). Or if we want to play with a new format for the response, that's fine too I guess.

If we're just adding new fields, we can probably keep it the same. It does feel like we're "versioning" it though just through the passing of a specific field. This is somewhat confusing, and we could think of a clearer way to explain this to users.

**Comment by stsewd** [comment]
> Both should work though, and we can combine them on our side?

I'd vote to have just one way to do it, I'm +1 on using the `key:value` syntax. Should we keep the same syntax for specifying a version? `project:docs:latest`, or use another separator for the version? `project:docs/latest`?

**Comment by stsewd** [comment]
> It does feel like we're "versioning" it though just through the passing of a specific field. This is somewhat confusing, and we could think of a clearer way to explain this to users.

yeah, maybe versioning could work, but If create a new version, I think we could think in expanding a little the response, for example have `project` be an object instead of a string with just the slug, that way we could combine these into a proper object (project.alias or version.project)

```
          "project": "docs",
          "project_alias": null,
```

we won't include the whole thing like https://docs.readthedocs.io/en/stable/api/v3.html#get--api-v3-projects-(string-project_slug)-, but just include the slug and leave the door open in case we need to another attributes.

Don't know, just a thought, maybe we don't need to expand those fields.

**Comment by humitos** [comment]
Meh, take that back 🙃 --GitHub and New Relic both use `key:value`; so it seems more standardized to follow that pattern as well. However, how do they handle complex attributes like `project:slug:version`? Probably, worth checking that to follow the same approach.

**Comment by stsewd** [comment]
This was implemented in the search API v3

**Pull request closed without merging**